File: rest/src/main/java/org/example/rest/resources/channel/message/AllowedMentions.java
package org.example.rest.resources.channel.message;

import java.util.List;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonValue;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;

import org.example.immutables.ImmutableJson;
import org.example.rest.resources.Snowflake;

@ImmutableJson
@JsonDeserialize(as = ImmutableAllowedMentions.class)
public interface AllowedMentions {

    static Builder builder() {
        return new Builder();
    }

    List<Type> parse();

    List<Snowflake> roles();

    List<Snowflake> users();

    @JsonProperty("replied_user")
    boolean repliedUser();

    enum Type {
        ROLE_MENTIONS("roles"),
        USER_MENTIONS("users"),
        EVERYONE_MENTIONS("everyone");

        private final String value;

        Type(String value) {
            this.value = value;
        }

        @JsonValue
        public String getValue() {
            return value;
        }
    }

    class Builder extends ImmutableAllowedMentions.Builder {
        protected Builder() {
        }
    }
}

File: rest/kotlin/src/main/kotlin/json/message/AllowedMentions.kt
package io.disquark.rest.kotlin.json.message

import io.disquark.rest.json.Snowflake
import io.disquark.rest.json.message.AllowedMentions as ImmutableAllowedMentions

data class AllowedMentions(
    var parse: List<ImmutableAllowedMentions.Type>? = null,
    var roles: List<Snowflake>? = null,
    var users: List<Snowflake>? = null,
    var repliedUser: Boolean? = null,
) {
    internal fun toImmutable() : ImmutableAllowedMentions {
        return ImmutableAllowedMentions.of()
            .run { parse?.let { withParse(it) } ?: this }
            .run { roles?.let { withRoles(it) } ?: this }
            .run { users?.let { withUsers(it) } ?: this }
            .run { repliedUser?.let { withRepliedUser(it) } ?: this }
    }
}
